### ISSUE TITLE
Segmentation fault on larger dataset using MaxPosteriorSampling

### ISSUE THREAD

**Issue by dong-tony** [pytorch/botorch#649]
## Issue description

This is part of the project I'm working on which I mentioned in Facebook/Ax#459. 

tldr i'm working on a chemistry problem where the search space is high dimensional (30+) and composed of discrete points (molecules). I'm still learning about this process so I'm experimenting with many approaches. One of the things I want to try is simply doing thompson sampling using `MaxPosteriorSampling` on my entire dataset (~49000 points), calculate the performance of the returned molecules (5 is the number i'm trying right now), and feed it back to the optimization loop manually. 

Everything runs fine when I was testing my code out on smaller subsets of the data, however when I tried running the full scale test I get a segmentation fault. I did a few more tests and seems like everything can still run up until ~42000 points, and after that it just throws the error again. Is this just an intrinsic limitation of the c code under the hood, or is there something I can do to solve this? Could it be that I'm running slightly older versions (see below)? I'm limited because i'm using a IBM cpu based computing cluster, which has the amount of memory I need to support this calculation, and seems like PyTorch 1.5 is the max they support currently.

Thanks and happy holidays!

## Code example

```
batch_size = 5

X = X_ini  # initial training set
Y = Y_ini
X_smiles = X_ini_smiles

train_Y = (Y - Y.mean()) / Y.std()

model = SingleTaskGP(X, train_Y)
mll = ExactMarginalLogLikelihood(model.likelihood, model)
fit_gpytorch_model(mll)

X_remain = torch.tensor(enu_features) # enu_features is a numpy array of shape (49000, 35)
thompson_sampling = MaxPosteriorSampling(model=model)
X_next = thompson_sampling(X_remain, num_samples=batch_size)
```

## System Info
- BoTorch Version 0.3.0
- GPyTorch Version 1.1.1
- PyTorch Version 1.5.0
- Linux RedHat 7.6

**Comment by saitcakmak** [comment]
I can't replicate the segmentation fault, but trying the same thing leads me to out of memory error. Segmentation fault maybe your CPUs way of complaining that it doesn't have enough memory. My rough estimate is that you'd need ~400GB memory to jointly sample from `49000` points. Do you need to jointly sample from that many points? If so, do you need to use "exact" sampling, or would approximate methods be fine?

Your code is trying to calculate the Cholesky decomposition of a `49000 x 49000` covariance matrix. This is probably where you're running into the segmentation fault. You could verify this by running the following (similar things happen within `MaxPosteriorSampling`):
```
import torch
from botorch.models import SingleTaskGP

batch_size = 5

model = SingleTaskGP(torch.rand(50, 35), torch.randn(50, 1))

X_remain = torch.rand(49000, 35)

# model.posterior(X_remain).rsample(sample_shape=torch.Size([batch_size]))

mvn = model.posterior(X_remain).mvn
covar = mvn.lazy_covariance_matrix
covar.cholesky()
```

**Comment by dong-tony** [comment]
Thanks for the suggestions! Hmm, the Cholesky decomp code actually works without any segmentation fault, I even bumped it up to 60000 points and it still works. Perhaps I should just try forking `MaxPosteriorSampling` and figuring out which operation within is causing the issue? 

I also would definitely like to learn about any more efficient ways of sampling jointly, do you have any specific suggestions for what I can try? Over at the Ax repo Eytan kindly suggested I use TuRBO-1 which definitely limits the number of joint samples for my case, but I was just curious to see if sampling everything would also work. Thanks!

**Comment by Balandat** [comment]
> I also would definitely like to learn about any more efficient ways of sampling jointly

So this is getting to a scale where the "vanilla" approach will hit limits. A few options: 
1. One could use approximate root decompositions instead of a full Cholesky decomposition. We'll have to add the ability to override a little option here: https://github.com/pytorch/botorch/blob/master/botorch/posteriors/gpytorch.py#L83 (we currently don't do that by default since non-determinism in the approximate computation causes challenges with BoTorch's SAA approach to optimizing acquisition functions). I can do that.
2. One could use the CIQ approach discussed in https://arxiv.org/pdf/2006.11267.pdf - we'd have to hook this up on the BoTorch end. @dme65 would be the right person for this :) 
3. One could also try using Random Fourier Feature approximations of the GP, but I would probably only try this if neither of the above worked out.

In general, one challenge will be that yo'd be constructing a `49000 x 49000`-dim dense test-test covariance. Is there some structure in the problem one could exploit? Like do you know upfront that certain subsets of the remaining points would be independent in some way?

**Comment by saitcakmak** [comment]
Interesting. I'd think the memory bottleneck would be the actual sampling from the posterior within `MaxPosteriorSampling`, which is the commented out line above and performs the Cholesky decomposition under the hood. This is where I run into OOM when replicating your code. 

FWIW, I myself ran into similar issues (namely illegal instruction error) when running code on a cluster and using old CPUs. My problem was resolved when I switched to newer machines. Hopefully yours doesn't come down to that.

For approximate methods, I am somewhat familiar with https://arxiv.org/abs/2002.09309 , which combines the inducing point approaches with random Fourier basis to approximately sample functions from GP posteriors. I implemented this to work with `SingleTaskGP` in https://github.com/saitcakmak/gp-sampling . I haven't tested this too much, so use at your own risk. But It should be pretty easy to use TS with it.

**Comment by dong-tony** [comment]
@saitcakmak 
Thanks, I'll double check whether it is just the cluster architecture causing the problem.

@Balandat 
How involved is implementing the options you mentioned? Upon second thought, although I'm quite personally interested in learning about all of this, it might take some time for me to grasp (i'm definitely more of a chemist than computer scientist) and for this particular project as long as the actual BO results (the molecules I get back) don't suffer I don't mind using less efficient methods since the function evaluations themselves would take way longer each iteration anyways - i'll just throw more nodes at it if I need :)

About the structure of the data, I don't believe there is any particular organization we can exploit - or I just can' tell :sweat_smile: - but the data itself is just 49000 molecules with 35 continuous descriptors each

I'll try to see if I can debug the segmentation fault and if that doesn't work I might just work with TuRBO-1 first since I have that implemented. Although one worry as I was experimenting with TuRBO is that since my data is discrete and they are not necessarily "evenly spread", alot of the data gets cropped out if I define a certain TR (say around the best point thus far), so i'm not sure if the whole dataset is being explored sufficiently. (I suppose thats what TuRBO-m is for?) Just wondering if you guys have any insights on this.

But anyways, thanks alot for compiling the list of suggestions! If not now, they will definitely come in handy later on when I have more time to explore!

**Comment by saitcakmak** [comment]
> One could use approximate root decompositions instead of a full Cholesky decomposition. We'll have to add the ability to override a little option here: https://github.com/pytorch/botorch/blob/master/botorch/posteriors/gpytorch.py#L83 

This one, you could very easily do locally. Find the `GPyTorchPosterior` source code in your environment, for example on a Mac, using a conda environment, mine is located at `/Users/saitcakmak/.conda/envs/botorch_test/lib/python3.8/site-packages/botorch/posteriors/gpytorch.py`. You just need to modify these lines:
https://github.com/pytorch/botorch/blob/4ad285eb4566e3b278ae484dde2a60bad79d0344/botorch/posteriors/gpytorch.py#L83-L86
Comment out line 83 and re-align lines 84-86 with the rest of the code. With this modification, you can wrap your code with GPyTorch context managers to use approximate root decomposition:
```
with gpytorch.settings.fast_computations():
    X_next = thompson_sampling(X_remain, num_samples=batch_size)
``` 
This should help with memory use. With 15k samples, it reduces the memory use from 11gb to 6gb for me.

**Comment by eytan** [comment]
Hi @dong-tony,

This is most definitely due to memory limits.  I know that @dme65 is working on preparing a tutorial on efficient Thompson sampling with RFFs and CIQ, but the ETA on that might be in a few weeks from now. @saitcakmak's code could also do the trick.

Re: your issue with data clumpiness + TuRBO—that's an interesting problem, and not something our team has come across in the past.  You could try some simple heuristic like doubling the width of the TuRBO search space until there are at least say, 1k points, or something like that.  For a 35d search space I would definitely recommend sticking with something like TuRBO, in which case you should hopefully have a lot fewer than 40k points to sample from.  TuRBO-m maintains m simultaneous search spaces and is most useful when your function evaluations are cheap, can be run parallel, and you plan on running a lot of them, so I am not sure if that's relevant for your use case.

FYI, if you are interested in learning more about RFFs and decoupled sampling, there are some nice blog posts that explain the intuition behind each of these methods:
http://gregorygundersen.com/blog/2019/12/23/random-fourier-features/
https://sml-group.cc/blog/2020-gp-sampling/

@saitcakmak, @Balandat, since there is a good deal of interest in TS, I wonder if it's worth upstreaming some of the decoupled sampling stuff at some point (also cc @wjmaddox as a heads up).

**Comment by dong-tony** [comment]
Hey @eytan, thanks again for all your suggestions and insights!

**Comment by dme65** [comment]
**Drawing larger samples using CIQ:** All of this is implemented in GPyTorch, so it should just be a matter of setting the right contexts to enable CIQ. It's possible that `covar_root_decomposition` being turned off may interfere with this so we may have to change that a tiny bit as @Balandat pointed out. I'm planning on putting up a large-scale Thompson sampling tutorial with CIQ as @eytan mentioned and I can try to take a stab at that in the next few days, but note that you will require a GPU and most likely also KeOps to sample on all molecules as the corresponding covariance matrix won't fit in memory.

**TuRBO-1**: In 35D, TuRBO should work much better than EI (I saw your post on the Ax Github page), so I'm interested in figuring out what is going on there. You want to make sure that you start with an initial trust region that is large enough to allow you to explore enough initially and doesn't cut away too much. One such option could be to pick the side length in each dimension so you can reach reach at least 50% of the values for each descriptor, probably even a bit more. You also want to make sure that you eventually shrink enough to allow TuRBO to converge, but doing so without cutting away too much in the end. One possible option could be to require that each descriptor can always reach at least one value in addition to the center, which will guarantee that the smallest possible trust region contains at least 2^35 points. Another option could be to put a minimum on the number of points in the trust region as suggested by @eytan.

**Comment by dong-tony** [comment]
Hey @dme65, thanks for your insights! About the TuRBO-1 results being identical to EI which I mentioned in the Ax repo: I was initially testing with EI as the AF wihin TuRBO on a very small subset of my data for which it was easy to just calculate the AF on all points and do `argmax`. Since EI on its own was able to figure out the global max fast within a few iterations, the effects of the TR did not really kick in. But for the full dataset I will definitely be utilizing TuRBO with batch TS and trying out the ideas you guys suggested!

**Comment by Balandat** [comment]
#652 allows to use the approximate root decomposition without having to make code changes - basically you'll just need to wrap the Posterior's `rsample` call into a `with gpytorch.settings.fast_computations(covar_root_decomposition=True):` context manager.

**Comment by dme65** [comment]
#654 (still WIP) shows how to do large-scale Thompson sampling using CIQ/RFFs.  There is some redundant TS code that was needed to work around the posterior call that #652 fixes. I'll clean that up once Max's PR is merged.

**Comment by Balandat** [comment]
Closing this as the tutorial should resolve this. Note that there was actually an issue with the implementation that meant that the CIQ option for sampling did not get used - this will be fixed in #701 and will be part of the imminent 0.4.0 release.
